feat(commands/canary): add TLS canary address display and improve architecture support

kgand opened the pull request:
Display the actual Thread Local Storage (TLS) address where the stack canary is stored,
eliminating the need for manual searching. This addresses and fixes issue https://github.com/pwndbg/pwndbg/issues/2511.

Key changes:
- Add architecture-specific TLS canary offsets with source references
- Implement robust TLS base address detection with fallback mechanisms
- Add support for x86_64 (fs:0x28), i386 (gs:0x14), and aarch64 (tpidr_el0+0x28)
- Add value verification to detect potential TLS address calculation issues
- Improve error handling and user feedback for unsupported architectures

The command now shows:
- AT_RANDOM location (seed value source)
- TLS address where canary is stored
- Actual canary value
- Stack locations where canary is copied

Example output:
pwndbg> canary
AT_RANDOM = 0x7fff05b46609 # points to global canary seed value
TLS Canary = 0x7fff05b46628 # address where canary is stored
Canary    = 0x989538cda22ba700 (may be incorrect on != glibc)

Closes https://github.com/pwndbg/pwndbg/issues/2511

kgand committed:
feat(commands/canary): add TLS canary address display and improve architecture support

Display the actual Thread Local Storage (TLS) address where the stack canary is stored,
eliminating the need for manual searching. This addresses issue #2511.

Key changes:
- Add architecture-specific TLS canary offsets with source references
- Implement robust TLS base address detection with fallback mechanisms
- Add support for x86_64 (fs:0x28), i386 (gs:0x14), and aarch64 (tpidr_el0+0x28)
- Add value verification to detect potential TLS address calculation issues
- Improve error handling and user feedback for unsupported architectures

The command now shows:
- AT_RANDOM location (seed value source)
- TLS address where canary is stored
- Actual canary value
- Stack locations where canary is copied

Fixes #2511

patryk4815 commented on pwndbg/commands/canary.py:
```diff
@@ -28,6 +46,36 @@ def canary_value():
+    arch = pwndbg.aglib.arch.name
+    
+    # Get TLS base address
+    tls_base = pwndbg.aglib.tls.find_address_with_register()
+    if not tls_base:
+        # Try pthread_self() as fallback
+        tls_base = pwndbg.aglib.tls.find_address_with_pthread_self()
+        if not tls_base:
+            return None
```
```suggestion
    tls_base = (
        pwndbg.aglib.tls.find_address_with_register()
        or pwndbg.aglib.tls.find_address_with_pthread_self()
    )
    if not tls_base:
        return None
```

kgand committed:
refactor(commands/canary): optimize TLS base address lookup logic

Improve the TLS base address lookup mechanism in find_tls_canary_addr() 
by leveraging Python's short-circuit evaluation with the 'or' operator.

Changes:
- Replace nested if-statements with a single expression using 'or'
- Maintain the same lookup priority (register-based first, then pthread_self)
- Improve code readability and maintainability
- Reduce code complexity while preserving functionality

Before:
    tls_base = pwndbg.aglib.tls.find_address_with_register()
    if not tls_base:
        tls_base = pwndbg.aglib.tls.find_address_with_pthread_self()
        if not tls_base:
            return None

After:
    tls_base = (
        pwndbg.aglib.tls.find_address_with_register()
        or pwndbg.aglib.tls.find_address_with_pthread_self()
    )
    if not tls_base:
        return None

This change follows Python's idiomatic approach to handling fallback values,
making the code more concise while maintaining its clarity and purpose.

Addresses review feedback from PR #2751.

Co-authored-by: patryk4815 <bux.patryk@gmail.com>

patryk4815 commented:
please fix lint, just run: `./lint.sh --fix` and commit/push

disconnect3d commented:
Ping @kgand can u update this? :)

disconnect3d commented:
Closing this due to inactivity.
